lib/portlayer/exec/exec.go

```diff
@@ -171,6 +182,147 @@ func eventCallback(ie events.Event) {
 	return
 }
 
+func waitForContainerStarted(container *Container) bool {
+	ctx, cancel := context.WithTimeout(context.Background(), propertyCollectorTimeout)
+	defer cancel()
+
+	if container.vm == nil {
+		log.Errorf("Container(%s) wait failed for vm is not found", container.ExecConfig.ID)
+		return false
+	}
+	log.Debugf("Waiting for container(%s) started", container.ExecConfig.ID)
+	for _, k := range extraconfig.CalculateKeys(container.ExecConfig, "Sessions.*.Detail.StartTime", "") {
+		var poweredoff bool
+		var detail string
+		waitFunc := func(pc []types.PropertyChange) bool {
+			for _, c := range pc {
+				if c.Op != types.PropertyChangeOpAssign {
+					continue
+				}
+
+				switch v := c.Val.(type) {
+				case types.ArrayOfOptionValue:
+					for _, value := range v.OptionValue {
+						// check the status of the key and return true if it's been set to non-nil
+						if k == value.GetOptionValue().Key {
+							detail = value.GetOptionValue().Value.(string)
+							if detail != "0" {
+								return true
+							}
+							break // continue the outer loop as we may have a powerState change too
+						}
+					}
+				case types.VirtualMachinePowerState:
+					if v != types.VirtualMachinePowerStatePoweredOn {
+						// Give up if the vm has powered off
+						poweredoff = true
+						return true
+					}
+				}
+
+			}
+			return false
+		}
+		log.Debugf("Waiting for container(%s) key %s", container.ExecConfig.ID, k)
+
+		err := container.vm.WaitForExtraConfig(ctx, waitFunc)
+		if poweredoff {
+			log.Errorf("Failed to wait container started for vm is powered off")
+			return false
+		}
+		if err != nil {
+			log.Errorf("Failed to wait container started: %s", err)
+		}
+	}
+	return true
+}
+
+func containerTimeCallback(ie events.Event) {
+	container := Containers.Container(ie.Reference())
+	if container == nil {
+		log.Debugf("Container(%s) is not found", ie.Reference())
+	}
+	switch ie.String() {
+	case events.ContainerStarted,
```

Comment:
Is there an existing callback that handles these events? If yes, why not put this code there?